working/macros/feature-specification.md

```diff
@@ -304,20 +304,69 @@ class, allowing you to build up your code fragments however you like.
 
 ## Scoping
 
-**TODO**: Fill in this section with more detail and a real proposal.
+### Resolved identifiers
 
 Macros will likely want to introduce references to identifiers that are not in
 the scope of the library in which they are running, but are in the scope of the
-macro itself.
+macro itself, or possibly even in some other library that is known to the
+macro, but not imported by the library where the macro is applied.
 
-We will want some way of providing an affordance to emit a reference to
-something from the macro scope, but in the code generated for the original
-library.
+Even if an identifier is expected to be in scope of the library in which the
+macro is running (lets say its exported by the macro library), that identifier
+could be shadowed by another identifier in the library.
 
-We do have the start of something like this already available in the `builder`
-api - these have APIs to get a reference to a `Type` object. We will want to add
-the ability to do the same for any arbitrary identifier, and then the ability
-to emit references to these inside of [Code][] objects.
+To enable a macro to safely emit a reference to a known identifier, there is
+a `ResolvedIdentifier` subtype of `Code`. This class takes both a simple name
+for the identifier (no prefix allowed), as well as a library uri, where that
+identifier should be looked up.
+
+The generated code should be equivalent to adding a new import to the library,
+with the specified uri and a unique prefix. In the code the identifier will be
+emitted with that unique prefix followed by its simple name.
+
+Note that technically this allows macros to add references to libraries that
+the macro itself does not depend on, and the users application also may not
+depend on. This is discouraged, but not prevented, and should result in an error
+if it happens.
+
+### Generated declarations
+
+Macros may add new declarations which collide with existing symbols in the
+library. We want to ensure that the intent of any user written code is always
+clear in this case. Consider the following example:
+
+```dart
+int get x => 1;
+
+@generateX 
+class Bar {
+  // Generated: int get x => 2;
+
+  // Should this return the top level `x`, or the generated instance getter?
+  int get y => x; 
+}
+```
+
+Every choice here has a lot of downsides, so it is instead an error. More
```

Comment:
It might not be clear to the reader what "choices" are being made. Maybe spell it out:

> We could say that any identifier that can be resolved before macro application keeps its original resolution. Or we could re-resolve all identifiers after the macros are applied, which can possibly change what they resolve to. We could make it some kind of error for a macro to introduce an identifier that shadows another. Or we could make it error to *use* an identifier shadowed by one produced by a macro.
>
> Every choice here has downsides...

Reply:
Reworked this a bunch